DSDEGP-2062: cram, arrays delivery refinements

pshapiro4broad opened the pull request:
### Description

https://broadinstitute.atlassian.net/browse/DSDEGP-2062

- refactored common delivery code into `DeliverExecutor`
- added check for existing workspace name in metadata to prevent re-delivery of already delivered data

----

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: Ensure that you have added or updated tests
- [x] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

pshapiro4broad committed:
DSDEGP-2062: refactor common delivery code into new class DeliveryExecutor

pshapiro4broad committed:
DSDEGP-2062: add check to prevent delivering a record that has already been delivered to a different workspace

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala:
```diff
@@ -0,0 +1,67 @@
+  ): Source[moveCommand.index.MetadataType, NotUsed] = {
+    val baseStream = super.checkPreconditions(ioUtil, webClient)
+    baseStream.flatMapConcat {
+      case (metadata) =>
```
`case` and parens not needed here.

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala:
```diff
@@ -0,0 +1,67 @@
+    val baseStream = super.checkPreconditions(ioUtil, webClient)
+    baseStream.flatMapConcat {
+      case (metadata) =>
+        if (metadata.workspaceName.isDefined && !metadata.workspaceName.get.equals(
```
I think this is `metadata.workspaceName.exists(_ != deliverCommand.workspaceName)`

pshapiro4broad replied:
Yeah, I had something a lot like that before but was still wrestling types at the time. I'll change it.

danxmoran commented on clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/WgsCramTests.scala:
```diff
@@ -1172,4 +1172,179 @@ trait WgsCramTests { self: BaseIntegrationSpec =>
     }
   }
+
+  it should "not allow a delivery if a record has already been delivered" in {
```
What problems did you hit trying to do this in the unit tests?

pshapiro4broad replied:
I was able to create a non-sealed base class for all commands, but it was hard to create the code to execute my test only command using the client main code. I have it shelved in git and I'll make another attempt on it tomorrow.

pshapiro4broad committed:
DSDEGP-2062: add "--force" option to delivery to override check for already delivered record

- other misc code review feedback

pshapiro4broad committed:
DSDEGP-2062: fix error in error message found while testing

pshapiro4broad committed:
DSDEGP-2062: implement tests using unit tests and mock, not integration tests

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala:
```diff
@@ -0,0 +1,67 @@
+    val baseStream = super.checkPreconditions(ioUtil, webClient)
+    baseStream.flatMapConcat { metadata =>
+      if (metadata.workspaceName.exists(
+            _ != deliverCommand.workspaceName && !deliverCommand.force
```
Would bust the `deliverCommand.force` out of & before the `exists`:
```scala
if (!deliverCommand.force && metadata.workspaceName.exists(...))
```

pshapiro4broad replied:
It depends on if you want to short-circuit the workspace name check or the force check. Doesn't matter to me either way.

danxmoran replied:
I think it'll just make for better formatting 😄 but the pattern we use elsewhere is for `force` to short-circuit even performing a check.

pshapiro4broad replied:
No such luck regarding the formatting
```
      if (deliverCommand.force || !metadata.workspaceName.exists(
            _ != deliverCommand.workspaceName
          )) {
```

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala:
```diff
@@ -20,68 +19,60 @@ import scala.concurrent.ExecutionContext
-                redIdatPath = Some(movedRedIdat)
-              )
+        val newMetadata = movedMetadata
+          .withWorkspaceName(deliverCommand.workspaceName)
```
Could we move the call to `withWorkspaceName` into the superclass?

pshapiro4broad replied:
OK

danxmoran commented on clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverWgsCramExecutor.scala:
```diff
@@ -22,37 +21,29 @@ import scala.concurrent.ExecutionContext
+      case (Some(cramMd5), Some(cramPath)) => {
+
+        val newMetadata =
+          movedMetadata.withWorkspaceName(deliverCommand.workspaceName)
```
Same question about moving this to the superclass

danxmoran commented on clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutorSpec.scala:
```diff
@@ -0,0 +1,159 @@
+  }
+
+  it should "deliver wgs cram with base name change" in {
+    successfulDelivery()
```
I think this is technically a feature of the move executor.

pshapiro4broad replied:
Sure, considering that the trivial delivery is a move.

danxmoran commented on clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutorSpec.scala:
```diff
@@ -0,0 +1,159 @@
+        ),
+        true
+      )
+      .returning(Source.single(id.asJson))
```
Do we need to duplicate most of this setup from the move executor, or can we just test the overridden methods?

pshapiro4broad committed:
Merge branch 'develop' into ps_DSDEGP-2062_cram_delivery_refinements

danxmoran commented on clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutorSpec.scala:
```diff
@@ -0,0 +1,159 @@
+    val (ioUtil: IoUtil, webClient: ClioWebClient) = initMocks(metadata)
+
+    val executor =
+      new DeliverWgsCramExecutor(
```
Using `DeliverWgsCramExecutor` brings along the baggage of the extra cram operations it defines. If we want to purely test the superclass, I'd make an anonymous instance with the abstract method overridden to be a no-op.

pshapiro4broad committed:
DSDEGP-2062: code review feedback

- check for `force` before checking workspace name
- move common `withWorkspaceName()` code into DeliverExecutor
- rewrite test code to only test new deliver precondition check code

pshapiro4broad committed:
Merge remote-tracking branch 'origin/ps_DSDEGP-2062_cram_delivery_refinements' into ps_DSDEGP-2062_cram_delivery_refinements

pshapiro4broad committed:
DSDEGP-2062: remove double negative to make it easier to read

pshapiro4broad committed:
Merge branch 'develop' into ps_DSDEGP-2062_cram_delivery_refinements
